### PR TITLE
[spmd] Use TupleStrategy and enable replicate fused_adam

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99964
* #99901
* #99900
* #99899
* #99898
* __->__ #99374
* #99489
* #99436
* #99897

### CODE REVIEW

File: torch/distributed/_spmd/data_parallel.py [hunk 1]
```diff
@@ -482,6 +484,38 @@ def build_data_parallel_strategies(
...
+                output_strategy_len = len(node.args) - 1
+                tuple_strategies = []
+                for i in range(output_strategy_len):
+                    if not isinstance(node.args[i], list):
+                        raise RuntimeError(
+                            f"Expecting list as arg to build Tuple Strategy, but found type {type(node.args[i])}!"
+                        )
+                    # for list/tuple arg, use the first one to find out the node type
+                    if len(node.args[i]) > 0:
+                        arg_strategy = dp_strategy_map[node.args[i][0]]
```
Comment (mrshenli) [inline torch/distributed/_spmd/data_parallel.py:502]: (not for this PR) do we need to redistribute all tensors in `args[i]` to make sure they are all replicated or sharded along the same dim?

File: torch/distributed/_spmd/data_parallel.py [hunk 2]
```diff
@@ -673,36 +707,39 @@ def mark_data_parallel_shardings(
...
+            return local_shard
+
+        for idx, placement in enumerate(spec.placements):
+            if placement.is_shard():
+                placement = cast(Shard, placement)
+                num_chunks = spec.mesh.size(dim=idx)
+                my_coord = spec.mesh.get_coordinate()
+                assert my_coord is not None, "current rank not in mesh!"
+                my_coord_on_mesh_dim = my_coord[idx]
+                local_shard = placement._split_tensor(
```
Comment (mrshenli) [inline torch/distributed/_spmd/data_parallel.py:728]: split tensor seems return a tuple of two elements?

https://github.com/pytorch/pytorch/blob/3afa60bf0f849dd61c49b66a937bb214bcd98f38/torch/distributed/_tensor/placement_types.py#L89-L91

File: torch/distributed/_spmd/data_parallel.py [hunk 3]
```diff
@@ -673,36 +707,39 @@ def mark_data_parallel_shardings(
...
-        # flag indicate the tensor is already local, we don't need to do anything
+    if isinstance(val, torch.Tensor):
+        local_shard = val
+        if val.ndim == 0:
+            # If it's already a scalar tensor, it is already local, we don't
+            # need to do anything
+            return local_shard
+
+        for idx, placement in enumerate(spec.placements):
+            if placement.is_shard():
```
Comment (mrshenli) [inline torch/distributed/_spmd/data_parallel.py:722]: does this work if the tensor is sharded across a 2D mesh?

**Review by mrshenli — APPROVED** [review]
stamp to unblock

File: torch/distributed/_spmd/data_parallel.py (hunk 3 above)
Comment (wanchaol) [reply torch/distributed/_spmd/data_parallel.py:722]: Yes, it should work automatically as we are looping over the placements

File: torch/distributed/_spmd/data_parallel.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_spmd/data_parallel.py:728]: Yep I used `[0]` at the end of get the `shard_list`

**Pull request closed without merging**